# Perfilado de datasets mineria
- Juan Pablo Mosquera
- David Zahid Riveros
- Oscar Manuel Contreras Gacha
- Anderson David Arenas Gutierrez

In [ ]:
import pandas as pd


#FIFA

In [ ]:
df_1 = pd.read_csv('fifa_eda_stats.csv')

In [ ]:
"""Limpieza de valores de dinero"""
def convert_value_to_numeric(value_str):
    if pd.isna(value_str) or not isinstance(value_str, str):
        return None
    value_str = value_str.replace('€', '').strip()
    if 'M' in value_str:
        return float(value_str.replace('M', '')) * 1_000_000
    elif 'K' in value_str:
        return float(value_str.replace('K', '')) * 1_000
    elif value_str == '0':
        return 0.0
    else:
        try:
            return float(value_str)
        except ValueError:
            return None
df_1['Value'] = df_1['Value'].apply(convert_value_to_numeric)
df_1['Wage'] = df_1['Wage'].apply(convert_value_to_numeric)
df_1['Release Clause'] = df_1['Release Clause'].apply(convert_value_to_numeric)

In [ ]:
print(df_1["Value"].value_counts(normalize=True).round(3))

Value
1100000.0     0.024
375000.0      0.020
425000.0      0.019
325000.0      0.019
450000.0      0.019
              ...  
63000000.0    0.000
51500000.0    0.000
53500000.0    0.000
58000000.0    0.000
81000000.0    0.000
Name: proportion, Length: 217, dtype: float64


In [ ]:
def perfilar(df):
    filas = []
    for col in df.columns:
        s = df[col]
        fila = {
            "columna":   col,
            "tipo":      str(s.dtype),
            "distintos": s.nunique(),
            "faltan_%":  round(s.isna().mean() * 100, 1),
            "moda":      s.mode().iloc[0] if not s.mode().empty else None,
        }
        if pd.api.types.is_numeric_dtype(s):
            fila["min"]     = round(s.min(), 2)
            fila["mediana"] = round(s.median(), 2)
            fila["media"]   = round(s.mean(), 2)
            fila["max"]     = round(s.max(), 2)
        filas.append(fila)
    return pd.DataFrame(filas).set_index("columna")

perfilar(df_1)


1. Corra `perfilar()` y pegue la salida.

<div id="df-812bf6ae-a487-4752-ad77-2c6b21f068fc" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>tipo</th>
      <th>distintos</th>
      <th>faltan_%</th>
      <th>moda</th>
      <th>min</th>
      <th>mediana</th>
      <th>media</th>
      <th>max</th>
    </tr>
    <tr>
      <th>columna</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>ID</th>
      <td>int64</td>
      <td>18207</td>
      <td>0.0</td>
      <td>16</td>
      <td>16.0</td>
      <td>221759.0</td>
      <td>214298.34</td>
      <td>246620.0</td>
    </tr>
    <tr>
      <th>Name</th>
      <td>object</td>
      <td>17194</td>
      <td>0.0</td>
      <td>J. Rodríguez</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Age</th>
      <td>int64</td>
      <td>29</td>
      <td>0.0</td>
      <td>21</td>
      <td>16.0</td>
      <td>25.0</td>
      <td>25.12</td>
      <td>45.0</td>
    </tr>
    <tr>
      <th>Nationality</th>
      <td>object</td>
      <td>164</td>
      <td>0.0</td>
      <td>England</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Overall</th>
      <td>int64</td>
      <td>48</td>
      <td>0.0</td>
      <td>66</td>
      <td>46.0</td>
      <td>66.0</td>
      <td>66.24</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>Potential</th>
      <td>int64</td>
      <td>47</td>
      <td>0.0</td>
      <td>70</td>
      <td>48.0</td>
      <td>71.0</td>
      <td>71.31</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>Club</th>
      <td>object</td>
      <td>651</td>
      <td>1.3</td>
      <td>AS Monaco</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Value</th>
      <td>object</td>
      <td>217</td>
      <td>0.0</td>
      <td>€1.1M</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Wage</th>
      <td>object</td>
      <td>144</td>
      <td>0.0</td>
      <td>€1K</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Preferred Foot</th>
      <td>object</td>
      <td>2</td>
      <td>0.3</td>
      <td>Right</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>International Reputation</th>
      <td>float64</td>
      <td>5</td>
      <td>0.3</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.11</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>Weak Foot</th>
      <td>float64</td>
      <td>5</td>
      <td>0.3</td>
      <td>3.0</td>
      <td>1.0</td>
      <td>3.0</td>
      <td>2.95</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>Skill Moves</th>
      <td>float64</td>
      <td>5</td>
      <td>0.3</td>
      <td>2.0</td>
      <td>1.0</td>
      <td>2.0</td>
      <td>2.36</td>
      <td>5.0</td>
    </tr>
    <tr>
      <th>Work Rate</th>
      <td>object</td>
      <td>9</td>
      <td>0.3</td>
      <td>Medium/ Medium</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Body Type</th>
      <td>object</td>
      <td>10</td>
      <td>0.3</td>
      <td>Normal</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Position</th>
      <td>object</td>
      <td>27</td>
      <td>0.3</td>
      <td>ST</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Jersey Number</th>
      <td>float64</td>
      <td>99</td>
      <td>0.3</td>
      <td>8.0</td>
      <td>1.0</td>
      <td>17.0</td>
      <td>19.55</td>
      <td>99.0</td>
    </tr>
    <tr>
      <th>Joined</th>
      <td>object</td>
      <td>1736</td>
      <td>8.5</td>
      <td>Jul 1, 2018</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Loaned From</th>
      <td>object</td>
      <td>341</td>
      <td>93.1</td>
      <td>Atalanta</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Contract Valid Until</th>
      <td>object</td>
      <td>36</td>
      <td>1.6</td>
      <td>2019</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Height</th>
      <td>object</td>
      <td>21</td>
      <td>0.3</td>
      <td>6'0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Weight</th>
      <td>object</td>
      <td>57</td>
      <td>0.3</td>
      <td>165lbs</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>Crossing</th>
      <td>float64</td>
      <td>89</td>
      <td>0.3</td>
      <td>62.0</td>
      <td>5.0</td>
      <td>54.0</td>
      <td>49.73</td>
      <td>93.0</td>
    </tr>
    <tr>
      <th>Finishing</th>
      <td>float64</td>
      <td>93</td>
      <td>0.3</td>
      <td>58.0</td>
      <td>2.0</td>
      <td>49.0</td>
      <td>45.55</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>HeadingAccuracy</th>
      <td>float64</td>
      <td>91</td>
      <td>0.3</td>
      <td>58.0</td>
      <td>4.0</td>
      <td>56.0</td>
      <td>52.30</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>ShortPassing</th>
      <td>float64</td>
      <td>85</td>
      <td>0.3</td>
      <td>64.0</td>
      <td>7.0</td>
      <td>62.0</td>
      <td>58.69</td>
      <td>93.0</td>
    </tr>
    <tr>
      <th>Volleys</th>
      <td>float64</td>
      <td>87</td>
      <td>0.3</td>
      <td>49.0</td>
      <td>4.0</td>
      <td>44.0</td>
      <td>42.91</td>
      <td>90.0</td>
    </tr>
    <tr>
      <th>Dribbling</th>
      <td>float64</td>
      <td>94</td>
      <td>0.3</td>
      <td>64.0</td>
      <td>4.0</td>
      <td>61.0</td>
      <td>55.37</td>
      <td>97.0</td>
    </tr>
    <tr>
      <th>Curve</th>
      <td>float64</td>
      <td>89</td>
      <td>0.3</td>
      <td>58.0</td>
      <td>6.0</td>
      <td>48.0</td>
      <td>47.17</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>FKAccuracy</th>
      <td>float64</td>
      <td>90</td>
      <td>0.3</td>
      <td>42.0</td>
      <td>3.0</td>
      <td>41.0</td>
      <td>42.86</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>LongPassing</th>
      <td>float64</td>
      <td>84</td>
      <td>0.3</td>
      <td>62.0</td>
      <td>9.0</td>
      <td>56.0</td>
      <td>52.71</td>
      <td>93.0</td>
    </tr>
    <tr>
      <th>BallControl</th>
      <td>float64</td>
      <td>90</td>
      <td>0.3</td>
      <td>65.0</td>
      <td>5.0</td>
      <td>63.0</td>
      <td>58.37</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Acceleration</th>
      <td>float64</td>
      <td>86</td>
      <td>0.3</td>
      <td>69.0</td>
      <td>12.0</td>
      <td>67.0</td>
      <td>64.61</td>
      <td>97.0</td>
    </tr>
    <tr>
      <th>SprintSpeed</th>
      <td>float64</td>
      <td>85</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>12.0</td>
      <td>67.0</td>
      <td>64.73</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Agility</th>
      <td>float64</td>
      <td>81</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>14.0</td>
      <td>66.0</td>
      <td>63.50</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Reactions</th>
      <td>float64</td>
      <td>68</td>
      <td>0.3</td>
      <td>65.0</td>
      <td>21.0</td>
      <td>62.0</td>
      <td>61.84</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Balance</th>
      <td>float64</td>
      <td>81</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>16.0</td>
      <td>66.0</td>
      <td>63.97</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>ShotPower</th>
      <td>float64</td>
      <td>92</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>2.0</td>
      <td>59.0</td>
      <td>55.46</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>Jumping</th>
      <td>float64</td>
      <td>74</td>
      <td>0.3</td>
      <td>70.0</td>
      <td>15.0</td>
      <td>66.0</td>
      <td>65.09</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>Stamina</th>
      <td>float64</td>
      <td>85</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>12.0</td>
      <td>66.0</td>
      <td>63.22</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Strength</th>
      <td>float64</td>
      <td>74</td>
      <td>0.3</td>
      <td>68.0</td>
      <td>17.0</td>
      <td>67.0</td>
      <td>65.31</td>
      <td>97.0</td>
    </tr>
    <tr>
      <th>LongShots</th>
      <td>float64</td>
      <td>92</td>
      <td>0.3</td>
      <td>59.0</td>
      <td>3.0</td>
      <td>51.0</td>
      <td>47.11</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>Aggression</th>
      <td>float64</td>
      <td>85</td>
      <td>0.3</td>
      <td>65.0</td>
      <td>11.0</td>
      <td>59.0</td>
      <td>55.87</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>Interceptions</th>
      <td>float64</td>
      <td>89</td>
      <td>0.3</td>
      <td>62.0</td>
      <td>3.0</td>
      <td>52.0</td>
      <td>46.70</td>
      <td>92.0</td>
    </tr>
    <tr>
      <th>Positioning</th>
      <td>float64</td>
      <td>94</td>
      <td>0.3</td>
      <td>58.0</td>
      <td>2.0</td>
      <td>55.0</td>
      <td>49.96</td>
      <td>95.0</td>
    </tr>
    <tr>
      <th>Vision</th>
      <td>float64</td>
      <td>85</td>
      <td>0.3</td>
      <td>58.0</td>
      <td>10.0</td>
      <td>55.0</td>
      <td>53.40</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>Penalties</th>
      <td>float64</td>
      <td>87</td>
      <td>0.3</td>
      <td>45.0</td>
      <td>5.0</td>
      <td>49.0</td>
      <td>48.55</td>
      <td>92.0</td>
    </tr>
    <tr>
      <th>Composure</th>
      <td>float64</td>
      <td>84</td>
      <td>0.3</td>
      <td>60.0</td>
      <td>3.0</td>
      <td>60.0</td>
      <td>58.65</td>
      <td>96.0</td>
    </tr>
    <tr>
      <th>Marking</th>
      <td>float64</td>
      <td>92</td>
      <td>0.3</td>
      <td>60.0</td>
      <td>3.0</td>
      <td>53.0</td>
      <td>47.28</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>StandingTackle</th>
      <td>float64</td>
      <td>90</td>
      <td>0.3</td>
      <td>66.0</td>
      <td>2.0</td>
      <td>55.0</td>
      <td>47.70</td>
      <td>93.0</td>
    </tr>
    <tr>
      <th>SlidingTackle</th>
      <td>float64</td>
      <td>88</td>
      <td>0.3</td>
      <td>62.0</td>
      <td>3.0</td>
      <td>52.0</td>
      <td>45.66</td>
      <td>91.0</td>
    </tr>
    <tr>
      <th>GKDiving</th>
      <td>float64</td>
      <td>71</td>
      <td>0.3</td>
      <td>8.0</td>
      <td>1.0</td>
      <td>11.0</td>
      <td>16.62</td>
      <td>90.0</td>
    </tr>
    <tr>
      <th>GKHandling</th>
      <td>float64</td>
      <td>70</td>
      <td>0.3</td>
      <td>10.0</td>
      <td>1.0</td>
      <td>11.0</td>
      <td>16.39</td>
      <td>92.0</td>
    </tr>
    <tr>
      <th>GKKicking</th>
      <td>float64</td>
      <td>79</td>
      <td>0.3</td>
      <td>12.0</td>
      <td>1.0</td>
      <td>11.0</td>
      <td>16.23</td>
      <td>91.0</td>
    </tr>
    <tr>
      <th>GKPositioning</th>
      <td>float64</td>
      <td>79</td>
      <td>0.3</td>
      <td>10.0</td>
      <td>1.0</td>
      <td>11.0</td>
      <td>16.39</td>
      <td>90.0</td>
    </tr>
    <tr>
      <th>GKReflexes</th>
      <td>float64</td>
      <td>76</td>
      <td>0.3</td>
      <td>10.0</td>
      <td>1.0</td>
      <td>11.0</td>
      <td>16.71</td>
      <td>94.0</td>
    </tr>
    <tr>
      <th>Release Clause</th>
      <td>object</td>
      <td>1244</td>
      <td>8.6</td>
      <td>€1.1M</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">
      
  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-812bf6ae-a487-4752-ad77-2c6b21f068fc')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">
      
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>
    
  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>
  </div>
  
  </div>

### Tres problemas encontrados y su solución

* **`Loaned From` con 93.1% de datos nulos:**  
  La gran mayoría de jugadores no están cedidos a préstamo, lo que genera una columna desierta de casi solo faltantes.  
  * **Solución:** Eliminar la columna por falta de varianza o convertirla en una bandera binaria `is_loaned` (1 si tiene club de origen, 0 si es `NaN`).

* **Valores numéricos guardados como texto (`Height` y `Weight`):**  
  Tienen formato de cadena tipo `6'0` y `165lbs`, lo que impide su análisis estadístico y cálculo de correlación directa.  
  * **Solución:** Aplicar una función de transformación en Pandas para convertir `Height` a centímetros y `Weight` a kilogramos (o libras numéricas) y castearlas a tipo `float64`.

* **Nulos masivos en cláusula de rescisión (`Release Clause` con 8.6% de faltantes):**  
  Existen 1,564 jugadores sin cláusula asignada en la base de datos.  
  * **Solución:** Imputar los faltantes calculando una estimación basada en la relación proporcional entre `Value` y `Overall`, o asignar 0 si el contrato no contempla dicha cláusula.

---

### Columna objetivo y su reparto (`Value`)

Al ser **`Value`** una variable continua económica (target de regresión), el método `value_counts(normalize=True)` se utiliza para observar la concentración del mercado o agrupar por rangos. Su reparto estadístico directo muestra:

* **Mínimo:** €0.0
* **Mediana:** €675,000.0
* **Media:** €2,410,695.89
* **Máximo:** €118,500,000.0

*(El reparto de `Value` presenta un marcado sesgo positivo a la derecha [right-skewed], donde más del 50% de los jugadores tiene un valor de mercado inferior a €675,000, mientras que una élite de superestrellas eleva la media a €2.4M y alcanza un pico de €118.5M).*

---

### ¿Este archivo aguanta el semestre?

**Sí, aguanta perfectamente el semestre**, porque cuenta con un volumen masivo de datos (18,207 registros), más de 80 variables con gran diversidad de tipos (métricas físicas, habilidades técnicas de 0 a 100, salarios y valor comercial) y múltiples desafíos de limpieza (nulos masivos, parseo de cadenas y transformaciones financieras) suficientes para todas las entregas de minería.

# Segundo dataset elegido NBA Players

In [ ]:
import pandas as pd
df_2 = pd.read_csv("/content/all_seasons.csv")

perfilar(df_2)

,tipo,distintos,faltan_%,moda,min,mediana,media,max
columna,,,,,,,,
Unnamed: 0,int64,12844,0.0,0,0.00,6421.50,6421.50,12843.00
player_name,object,2551,0.0,Vince Carter,NaN,NaN,NaN,NaN
team_abbreviation,object,36,0.0,CLE,NaN,NaN,NaN,NaN
age,float64,27,0.0,24.0,18.00,26.00,27.05,44.00
player_height,float64,30,0.0,205.74,160.02,200.66,200.56,231.14
player_weight,float64,157,0.0,99.79024,60.33,99.79,100.26,163.29
college,object,356,14.4,Kentucky,NaN,NaN,NaN,NaN
country,object,82,0.0,USA,NaN,NaN,NaN,NaN
draft_year,object,48,0.0,Undrafted,NaN,NaN,NaN,NaN


1. Corra `perfilar()` y pegue la salida.

<div id="df-8d7e1ccf-4eed-4385-aaeb-1daff7f55577" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>tipo</th>
      <th>distintos</th>
      <th>faltan_%</th>
      <th>moda</th>
      <th>min</th>
      <th>mediana</th>
      <th>media</th>
      <th>max</th>
    </tr>
    <tr>
      <th>columna</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Unnamed: 0</th>
      <td>int64</td>
      <td>12844</td>
      <td>0.0</td>
      <td>0</td>
      <td>0.00</td>
      <td>6421.50</td>
      <td>6421.50</td>
      <td>12843.00</td>
    </tr>
    <tr>
      <th>player_name</th>
      <td>object</td>
      <td>2551</td>
      <td>0.0</td>
      <td>Vince Carter</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>team_abbreviation</th>
      <td>object</td>
      <td>36</td>
      <td>0.0</td>
      <td>CLE</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>age</th>
      <td>float64</td>
      <td>27</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>18.00</td>
      <td>26.00</td>
      <td>27.05</td>
      <td>44.00</td>
    </tr>
    <tr>
      <th>player_height</th>
      <td>float64</td>
      <td>30</td>
      <td>0.0</td>
      <td>205.74</td>
      <td>160.02</td>
      <td>200.66</td>
      <td>200.56</td>
      <td>231.14</td>
    </tr>
    <tr>
      <th>player_weight</th>
      <td>float64</td>
      <td>157</td>
      <td>0.0</td>
      <td>99.79024</td>
      <td>60.33</td>
      <td>99.79</td>
      <td>100.26</td>
      <td>163.29</td>
    </tr>
    <tr>
      <th>college</th>
      <td>object</td>
      <td>356</td>
      <td>14.4</td>
      <td>Kentucky</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>country</th>
      <td>object</td>
      <td>82</td>
      <td>0.0</td>
      <td>USA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>draft_year</th>
      <td>object</td>
      <td>48</td>
      <td>0.0</td>
      <td>Undrafted</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>draft_round</th>
      <td>object</td>
      <td>9</td>
      <td>0.0</td>
      <td>1</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>draft_number</th>
      <td>object</td>
      <td>76</td>
      <td>0.0</td>
      <td>Undrafted</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>gp</th>
      <td>int64</td>
      <td>85</td>
      <td>0.0</td>
      <td>82</td>
      <td>1.00</td>
      <td>57.00</td>
      <td>51.15</td>
      <td>85.00</td>
    </tr>
    <tr>
      <th>pts</th>
      <td>float64</td>
      <td>319</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.00</td>
      <td>6.70</td>
      <td>8.21</td>
      <td>36.10</td>
    </tr>
    <tr>
      <th>reb</th>
      <td>float64</td>
      <td>153</td>
      <td>0.0</td>
      <td>2.0</td>
      <td>0.00</td>
      <td>3.00</td>
      <td>3.56</td>
      <td>16.30</td>
    </tr>
    <tr>
      <th>ast</th>
      <td>float64</td>
      <td>115</td>
      <td>0.0</td>
      <td>0.3</td>
      <td>0.00</td>
      <td>1.20</td>
      <td>1.82</td>
      <td>11.70</td>
    </tr>
    <tr>
      <th>net_rating</th>
      <td>float64</td>
      <td>772</td>
      <td>0.0</td>
      <td>-0.3</td>
      <td>-250.00</td>
      <td>-1.30</td>
      <td>-2.23</td>
      <td>300.00</td>
    </tr>
    <tr>
      <th>oreb_pct</th>
      <td>float64</td>
      <td>216</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>0.04</td>
      <td>0.05</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>dreb_pct</th>
      <td>float64</td>
      <td>352</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>0.13</td>
      <td>0.14</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>usg_pct</th>
      <td>float64</td>
      <td>348</td>
      <td>0.0</td>
      <td>0.183</td>
      <td>0.00</td>
      <td>0.18</td>
      <td>0.18</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>ts_pct</th>
      <td>float64</td>
      <td>563</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>0.53</td>
      <td>0.51</td>
      <td>1.50</td>
    </tr>
    <tr>
      <th>ast_pct</th>
      <td>float64</td>
      <td>483</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>0.10</td>
      <td>0.13</td>
      <td>1.00</td>
    </tr>
    <tr>
      <th>season</th>
      <td>object</td>
      <td>27</td>
      <td>0.0</td>
      <td>2021-22</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">
      
  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-8d7e1ccf-4eed-4385-aaeb-1daff7f55577')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">
      
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>
    
  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>
  </div>
  </div>

2. Escriba **tres problemas** que encontró y qué haría con cada uno.
   1. El primer problema que se encontró fue la columna 'unnamed: 0' que representa una fuga de datos y es un indice redundante pues es un indice autogenerado. Se soluciona eliminando directamente la columna con `df.drop`.
   2. Existen valores nullos en 'college' (14.4% de faltantes), existen jugadores que fueron reclutados de la secundaria o internacionalmente. La solución en este caso es pasar de 'NaN' a 'None' o 'International/HighSchool' para no perder esa información categórica.
   3. En las variables que destacan al draft como tal, 'draft_year', 'draft_round', 'draft_number'. Se leen como 'object' debido a que existe el texto 'undrafted', esto implide su tratamiento númerico o análisis de correlación directo. La solución en este caso seria crear una bandera binaria `is_drafted` (1 si si, 0 si no) y convertir 'Undrafted' a 0 en las columnas numéricas para que estas se puedan convertir a int / float.

3. Diga cuál es la **columna objetivo** y su reparto.
La columna objetivo es 'pts'. Es una variable continua, su target es de regresión. El reparto se describe mediante distribución continua:
- Mínimo: 0.00 pts
- Mediana: 6.70 pts
- Media: 8.21 pts
- Máximo: 36.10 pts

Este reparto presenta un sesgo positivo hacia la derecha, donde el 50% de los registros anota 6.70 puntos o menos. Una minoría elitista de anotadores eleva la media a 8.21 pts y alcanza un pico en 36.10 pts.

4. Una frase: ¿este archivo aguanta el semestre, sí o no, y por qué?
Sí aguanta perfectamente el semestre porque cuenta con un volumen sólido de datos (12,844 registros), 22 variables con una mezcla rica entre tipos numéricos y categóricos, una variable objetivo clara 'pts' para modelos de regresión, suficientes desafíos de preprocesamiento (nulos, tipados de caracteres y detección de atípicos) para justificar las fases de minería.

# Tercer Dataset elegido IMDB Movies

In [ ]:
import pandas as pd
df_3 = pd.read_csv("/content/movie_metadata.csv")

In [ ]:
perfilar(df_3)

,tipo,distintos,faltan_%,moda,min,mediana,media,max
columna,,,,,,,,
color,object,2,0.4,Color,NaN,NaN,NaN,NaN
director_name,object,2398,2.1,Steven Spielberg,NaN,NaN,NaN,NaN
num_critic_for_reviews,float64,528,1.0,1.0,1.00,110.00,140.19,8.130000e+02
duration,float64,191,0.3,90.0,7.00,103.00,107.20,5.110000e+02
director_facebook_likes,float64,435,2.1,0.0,0.00,49.00,686.51,2.300000e+04
actor_3_facebook_likes,float64,906,0.5,1000.0,0.00,371.50,645.01,2.300000e+04
actor_2_name,object,3032,0.3,Morgan Freeman,NaN,NaN,NaN,NaN
actor_1_facebook_likes,float64,878,0.1,1000.0,0.00,988.00,6560.05,6.400000e+05
gross,float64,4035,17.5,3000000.0,162.00,25517500.00,48468407.53,7.605058e+08


1. Corra `perfilar()` y pegue la salida.

<div id="df-511615cf-07dd-4ccf-8337-24c53623afd3" class="colab-df-container">
    <div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>tipo</th>
      <th>distintos</th>
      <th>faltan_%</th>
      <th>moda</th>
      <th>min</th>
      <th>mediana</th>
      <th>media</th>
      <th>max</th>
    </tr>
    <tr>
      <th>columna</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>color</th>
      <td>object</td>
      <td>2</td>
      <td>0.4</td>
      <td>Color</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>director_name</th>
      <td>object</td>
      <td>2398</td>
      <td>2.1</td>
      <td>Steven Spielberg</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>num_critic_for_reviews</th>
      <td>float64</td>
      <td>528</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.00</td>
      <td>110.00</td>
      <td>140.19</td>
      <td>8.130000e+02</td>
    </tr>
    <tr>
      <th>duration</th>
      <td>float64</td>
      <td>191</td>
      <td>0.3</td>
      <td>90.0</td>
      <td>7.00</td>
      <td>103.00</td>
      <td>107.20</td>
      <td>5.110000e+02</td>
    </tr>
    <tr>
      <th>director_facebook_likes</th>
      <td>float64</td>
      <td>435</td>
      <td>2.1</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>49.00</td>
      <td>686.51</td>
      <td>2.300000e+04</td>
    </tr>
    <tr>
      <th>actor_3_facebook_likes</th>
      <td>float64</td>
      <td>906</td>
      <td>0.5</td>
      <td>1000.0</td>
      <td>0.00</td>
      <td>371.50</td>
      <td>645.01</td>
      <td>2.300000e+04</td>
    </tr>
    <tr>
      <th>actor_2_name</th>
      <td>object</td>
      <td>3032</td>
      <td>0.3</td>
      <td>Morgan Freeman</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>actor_1_facebook_likes</th>
      <td>float64</td>
      <td>878</td>
      <td>0.1</td>
      <td>1000.0</td>
      <td>0.00</td>
      <td>988.00</td>
      <td>6560.05</td>
      <td>6.400000e+05</td>
    </tr>
    <tr>
      <th>gross</th>
      <td>float64</td>
      <td>4035</td>
      <td>17.5</td>
      <td>3000000.0</td>
      <td>162.00</td>
      <td>25517500.00</td>
      <td>48468407.53</td>
      <td>7.605058e+08</td>
    </tr>
    <tr>
      <th>genres</th>
      <td>object</td>
      <td>914</td>
      <td>0.0</td>
      <td>Drama</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>actor_1_name</th>
      <td>object</td>
      <td>2097</td>
      <td>0.1</td>
      <td>Robert De Niro</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>movie_title</th>
      <td>object</td>
      <td>4917</td>
      <td>0.0</td>
      <td>Ben-Hur</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>num_voted_users</th>
      <td>int64</td>
      <td>4826</td>
      <td>0.0</td>
      <td>57</td>
      <td>5.00</td>
      <td>34359.00</td>
      <td>83668.16</td>
      <td>1.689764e+06</td>
    </tr>
    <tr>
      <th>cast_total_facebook_likes</th>
      <td>int64</td>
      <td>3978</td>
      <td>0.0</td>
      <td>0</td>
      <td>0.00</td>
      <td>3090.00</td>
      <td>9699.06</td>
      <td>6.567300e+05</td>
    </tr>
    <tr>
      <th>actor_3_name</th>
      <td>object</td>
      <td>3521</td>
      <td>0.5</td>
      <td>Ben Mendelsohn</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>facenumber_in_poster</th>
      <td>float64</td>
      <td>19</td>
      <td>0.3</td>
      <td>0.0</td>
      <td>0.00</td>
      <td>1.00</td>
      <td>1.37</td>
      <td>4.300000e+01</td>
    </tr>
    <tr>
      <th>plot_keywords</th>
      <td>object</td>
      <td>4760</td>
      <td>3.0</td>
      <td>based on novel</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>movie_imdb_link</th>
      <td>object</td>
      <td>4919</td>
      <td>0.0</td>
      <td>http://www.imdb.com/title/tt0077651/?ref_=fn_t...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>num_user_for_reviews</th>
      <td>float64</td>
      <td>954</td>
      <td>0.4</td>
      <td>1.0</td>
      <td>1.00</td>
      <td>156.00</td>
      <td>272.77</td>
      <td>5.060000e+03</td>
    </tr>
    <tr>
      <th>language</th>
      <td>object</td>
      <td>46</td>
      <td>0.3</td>
      <td>English</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>country</th>
      <td>object</td>
      <td>65</td>
      <td>0.1</td>
      <td>USA</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>content_rating</th>
      <td>object</td>
      <td>18</td>
      <td>6.0</td>
      <td>R</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>budget</th>
      <td>float64</td>
      <td>439</td>
      <td>9.8</td>
      <td>20000000.0</td>
      <td>218.00</td>
      <td>20000000.00</td>
      <td>39752620.44</td>
      <td>1.221550e+10</td>
    </tr>
    <tr>
      <th>title_year</th>
      <td>float64</td>
      <td>91</td>
      <td>2.1</td>
      <td>2009.0</td>
      <td>1916.00</td>
      <td>2005.00</td>
      <td>2002.47</td>
      <td>2.016000e+03</td>
    </tr>
    <tr>
      <th>actor_2_facebook_likes</th>
      <td>float64</td>
      <td>917</td>
      <td>0.3</td>
      <td>1000.0</td>
      <td>0.00</td>
      <td>595.00</td>
      <td>1651.75</td>
      <td>1.370000e+05</td>
    </tr>
    <tr>
      <th>imdb_score</th>
      <td>float64</td>
      <td>78</td>
      <td>0.0</td>
      <td>6.7</td>
      <td>1.60</td>
      <td>6.60</td>
      <td>6.44</td>
      <td>9.500000e+00</td>
    </tr>
    <tr>
      <th>aspect_ratio</th>
      <td>float64</td>
      <td>22</td>
      <td>6.5</td>
      <td>2.35</td>
      <td>1.18</td>
      <td>2.35</td>
      <td>2.22</td>
      <td>1.600000e+01</td>
    </tr>
    <tr>
      <th>movie_facebook_likes</th>
      <td>int64</td>
      <td>876</td>
      <td>0.0</td>
      <td>0</td>
      <td>0.00</td>
      <td>166.00</td>
      <td>7525.96</td>
      <td>3.490000e+05</td>
    </tr>
  </tbody>
</table>
</div>
    <div class="colab-df-buttons">
      
  <div class="colab-df-container">
    <button class="colab-df-convert" onclick="convertToInteractive('df-511615cf-07dd-4ccf-8337-24c53623afd3')"
            title="Convert this dataframe to an interactive table."
            style="display:none;">
      
  <svg xmlns="http://www.w3.org/2000/svg" height="24px" viewBox="0 -960 960 960">
    <path d="M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z"/>
  </svg>
    </button>
    
  <style>
    .colab-df-container {
      display:flex;
      gap: 12px;
    }

    .colab-df-convert {
      background-color: #E8F0FE;
      border: none;
      border-radius: 50%;
      cursor: pointer;
      display: none;
      fill: #1967D2;
      height: 32px;
      padding: 0 0 0 0;
      width: 32px;
    }

    .colab-df-convert:hover {
      background-color: #E2EBFA;
      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);
      fill: #174EA6;
    }

    .colab-df-buttons div {
      margin-bottom: 4px;
    }

    [theme=dark] .colab-df-convert {
      background-color: #3B4455;
      fill: #D2E3FC;
    }

    [theme=dark] .colab-df-convert:hover {
      background-color: #434B5C;
      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);
      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));
      fill: #FFFFFF;
    }
  </style>
  </div>
  
  </div>

2. Escriba **tres problemas** que encontró y qué haría con cada uno.
   1. `gross` y `budget` tienen un porcentaje alto de faltantes (17.5% y 9.8%) y además presentan valores atípicos extremos vía IQR (318 atípicos en `budget`, 306 en `gross`); el máximo de `budget` llega a más de 12,000 millones de dólares, una cifra imposible para el presupuesto real de una sola película, lo que sugiere registros en monedas distintas al USD sin convertir. La solución sería primero investigar y corregir esos presupuestos (convertir a USD o marcarlos/eliminarlos si no se puede verificar la moneda), y para los faltantes crear una bandera binaria `gross_faltante` / `budget_faltante` antes de imputar, ya que la ausencia no es aleatoria (ver punto 3).
   2. El dataset tiene **45 filas duplicadas**, lo que puede sesgar las estadísticas descriptivas y sobre-representar ciertas películas en el análisis o en el entrenamiento de un modelo. La solución es eliminarlas con `df.drop_duplicates()`, revisando antes si son duplicados exactos en todas las columnas o solo coinciden en `movie_title` con datos ligeramente distintos (para no borrar películas legítimamente repetidas, como remakes).
   3. La columna `color` es casi constante (solo 2 valores únicos, dominada por "Color") por lo que aporta poca información predictiva por sí sola. Sin embargo, se descubrió que el porcentaje de faltantes en `gross` cambia fuertemente según `color` (33.5% en "Black and White" vs 16.6% en "Color") y según `content_rating` (de 4.2% en PG-13 hasta 100% en varias categorías de TV), es decir, la ausencia de `gross` no es aleatoria (MNAR). La solución no es eliminar `color` sin más, sino conservarla como variable auxiliar/indicador de faltante informativo en vez de imputar `gross` de forma ciega.
   
3. Diga cuál es la **columna objetivo** y su reparto.

La columna objetivo es `imdb_score`. Es una variable continua, por lo que el problema es de **regresión**. Su reparto, según `df["imdb_score"].describe()`:
- Mínimo: 1.60
- Percentil 25%: 5.80
- Mediana (50%): 6.60
- Percentil 75%: 7.20
- Máximo: 9.50
- Media: 6.44 | Desviación estándar: 1.13

A diferencia de una variable con sesgo marcado hacia un extremo, aquí el reparto es bastante simétrico y ligeramente sesgado a la izquierda (la media, 6.44, es apenas menor que la mediana, 6.60), con la mayoría de las películas concentradas entre 5.8 y 7.2 puntos, y colas cortas hacia calificaciones muy bajas o muy altas.

4. Una frase: ¿este archivo aguanta el semestre, sí o no, y por qué?

Sí aguanta el semestre, porque tiene un volumen considerable de datos (5,043 registros, 28 variables con mezcla de tipos: 13 float, 12 object, 3 int), una variable objetivo continua clara (`imdb_score`) para un problema de regresión, y suficientes retos reales de preprocesamiento —faltantes de hasta 17.5% con patrón no aleatorio (MNAR), 45 duplicados, valores atípicos extremos en `budget` y `gross`, alta cardinalidad en columnas de texto y una variable casi constante— como para justificar de sobra las distintas fases de minería de datos.